### ISSUE TITLE
[FIX-022] Distinguish `None` from zero in dataset batch-size resolution

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#232]
**Category:** Data contracts · **Priority:** P2 · **Size:** S

**Evidence status:** Reproduced input hardening.

## Outcome and scope

Use `None` exclusively as the automatic full-split batch-size sentinel in vision, tabular, preference and neighbor-graph wrappers. Reject malformed tuples and invalid explicit sizes before external dataset construction/tokenization where possible. Preserve each wrapper's existing empty-split reporting conventions and full-graph prohibition of explicit sizes.

## Current behavior / evidence

All four wrappers use `requested_size or split_size`. Consequently zero and False silently become full-batch requests, changing memory demand and optimization cadence. Tabular/preference resolve absent optional splits to a placeholder batch size of one but publish loader None; graph empty-split behavior is separately repaired by FIX-019. Vision test-loader behavior must not be changed as an incidental cleanup.

## Repository map

- [`src/nnx/nn/dataset/nn_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_dataset.py#L35) — Validate before invoking the dataset factory or accessing samples.
- [`src/nnx/nn/dataset/nn_tabular_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_tabular_dataset.py#L81) — Validate before tensor conversion/splitting; preserve optional loader conventions.
- [`src/nnx/nn/dataset/nn_preference_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_preference_dataset.py#L111) — Validate before tokenizer.encode calls.
- [`src/nnx/nn/dataset/nn_graph_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_graph_dataset.py#L82) — Neighbor resolution and full-mode explicit-size rejection.
- [`tests/test_nn_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_nn_dataset.py#L49) — Offline vision factory fixtures.
- [`tests/test_nn_graph_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_nn_graph_dataset.py#L31) — Offline neighbor construction fixture.
- [`tests/test_pass2_f_series.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_pass2_f_series.py#L1113) — Existing tabular and preference data contract coverage.

## Implementation approach

1. Add one private helper, proposed `src/nnx/nn/dataset/_validation.py`, validating a three-element tuple of None or positive non-boolean `numbers.Integral`. Normalize supported integer scalars to Python int, consistent with FIX-021. Reject lists/other containers with a clear tuple-contract error rather than silently accepting ambiguous shapes; reject strings and fractional/integer-valued floats.
2. Call validation first in each wrapper, before any dataset factory, tokenizer or RNG-consuming split. Name the wrapper and offending split/index, for example `batch_sizes[0] (train) must be None or a positive integer`. Use `ValueError` for malformed length/value/shape.
3. Replace truthiness resolution with `split_default if requested is None else requested`. Preserve the tabular/preference/vision `max(1, empty_count)` placeholders and optional loader decisions; a valid explicit positive value remains unchanged even when a split is empty. FIX-019 alone establishes graph absence reporting; coordinate this one derived-value exception with its author.
4. Full graph mode must reject every explicit positive size after validation, retaining the existing explanatory message. Do not accept an explicit zero as a way to disable a split. Prove rejection occurs before user factories are invoked using a call counter or raising factory.

## Acceptance criteria

- [ ] Zero, negative, fractional and boolean sizes fail consistently with split/index context in vision, tabular, preference and neighbor-graph wrappers.
- [ ] None continues to resolve to full split sizes, preserving valid existing results.
- [ ] Positive integer sizes are preserved verbatim; malformed tuple lengths fail with ValueError rather than an index error after dataset work.
- [ ] Existing empty-validation/test split conventions remain intact, and sampler='full' still rejects explicit batch sizes.

## Validation and completion

Minimal public reproduction:

```python
import pandas as pd
import pytest
from nnx import NNTabularDataset

frame = pd.DataFrame({"x": [1.0, 2.0], "y": [0, 1]})
kwargs = dict(df=frame, feature_cols=["x"], target_col="y",
              val_proportion=0.0, test_proportion=0.0)
with pytest.raises(ValueError, match=r"batch_sizes\[0\]"):
    NNTabularDataset(**kwargs, batch_sizes=(0, None, None))
ds = NNTabularDataset(**kwargs, batch_sizes=(None, None, None))
assert ds.batch_sizes == (2, 1, 1) and ds.val_loader is None
```

Proposed `tests/test_dataset_batch_sizes.py` should parameterize the four wrappers and all three positions over 0, -1, .5, True, False and malformed tuple lengths. Use a counting tokenizer for preferences and local factories for vision/graphs. Verify positive size 7 remains 7 rather than being clipped to a two-row split.

Existing coverage runs with `uv run --frozen pytest tests/test_nn_dataset.py tests/test_nn_graph_dataset.py tests/test_nn_graph_dataset_full_batch.py tests/test_pass2_f_series.py`. After adding the new file, run `uv run --frozen pytest tests/test_dataset_batch_sizes.py`. Neighbor checks need PyG import but no sampler iteration/download; preference tests need only their local tokenizer stub.

From the repository root, CI-equivalent setup is `uv sync --frozen --all-extras`. The commands above describe required implementation checks; none were executed during ticket enrichment. Finish with `uv run --frozen ruff check src/ tests/ examples/ scripts/`, `uv run --frozen ruff format --check src/ tests/ examples/ scripts/`, and `uv run --frozen pyright --warnings`. Preserve Python 3.10 compatibility and default-omission serialization; update canonical documentation sources, not generated projections by hand.

## Cascading changes and affected consumers

- **Shared input boundary with distinct outputs.** All four dataset wrappers publish resolved `batch_sizes`, but their empty-split conventions differ. Preserve the vision/tabular/preference placeholder-one metadata and graph's absent-split zero policy from [FIX-019](https://github.com/thekaveh/NNx/issues/229). The [vision wrapper](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_dataset.py#L35) must reject invalid requests before calling user dataset factories; the [preference wrapper](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_preference_dataset.py#L111) must do so before tokenizer.encode. Full-graph mode continues rejecting every explicit size, including otherwise valid positive values.
- **Training and workload consequences.** None still resolves to the full valid split and a positive value larger than the split still yields a smaller final batch; do not clip the published requested value or silently change optimizer cadence. Exercise a valid tabular wrapper through a short train call and compare batch/update counts against the resolved loader. Metadata/table output must agree with the loader's presence without conflating a derived zero with the forbidden explicit zero. No batch-size request becomes part of NNRun identity indirectly; existing data_id ownership stays caller-controlled.
- **Samples and documentation.** The [DPO sample](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/22_dpo_synthetic_preferences.py#L138) and tabular quickstart are concrete consumers. Add a small preference-construction helper that accepts a stub tokenizer and size tuple; validate rejected values without remote or tokenization side effects. Execute valid positive and automatic-size cases using the synthetic triples. Explain that None means automatic full split and zero does not disable validation. Preserve import-without-main behavior and explicitly register execution of the helper; existing discovery only imports sample22.

### Samples and executable coverage

- [`examples/22_dpo_synthetic_preferences.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/22_dpo_synthetic_preferences.py): Factor dataset construction into an offline helper accepting tokenizer/triples/batch_sizes; retain real local BPE main. **Verify:** Proposed test_dpo_sample_batch_sizes uses a counting stub tokenizer, proves invalid tuples make zero encode calls, and checks None versus positive sizes produce expected batch counts; explicit execution registration required.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/quickstart.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/quickstart.md) | `generated/site/Quickstart.md` | `generated/wiki/Quickstart.md` | Define None versus positive tuple entries, forbid zero/False, and explain optional empty-split loader behavior in the tabular example. |
| [`docs/dpo.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/dpo.md) | `generated/site/DPO.md` | `generated/wiki/DPO.md` | Document the exact batch_sizes tuple contract beside the preference dataset example. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] Invalid tuples passed through the vision factory and preference sample helper fail before dataset construction or tokenizer encoding, verified by zero call counts.
- [ ] For a valid two-row split, requested size seven remains seven in dataset metadata while iteration yields one two-row batch; None resolves to two.
- [ ] The bounded synthetic DPO sample execution exercises automatic and positive batch sizes and explicitly rejects zero as a split-disabling mechanism.
- [ ] The issue-specific changes in `docs/quickstart.md`, `docs/dpo.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** The traced downstream changes belong in this ticket. Shared-file coordination and optional combinations do not add a hard prerequisite.

Additional nonblocking consumers/coordination: [FIX-019](https://github.com/thekaveh/NNx/issues/229), [FIX-021](https://github.com/thekaveh/NNx/issues/231).

## Delivery order and dependencies

- **Execution order:** 33 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FIX-019](https://github.com/thekaveh/NNx/issues/229), [FIX-021](https://github.com/thekaveh/NNx/issues/231). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. Coordinate with FIX-019/021; preserve the documented None sentinel.

**Related existing issues (distinct scope):** [#188](https://github.com/thekaveh/NNx/issues/188).

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved in #301 (develop, merged `3d79994`) and #302 (main, merged `c518de6`). Folds the #188 documentation request.

**Root cause.** All four wrappers resolved `requested or split_size`, so `0` / `False` silently became full-split requests (different memory demand and optimizer cadence), a list or short tuple failed with `IndexError` only after the dataset had been built, and fractional / boolean / NumPy-integer entries were rejected — or not — deep inside `DataLoader` with no field context.

**What changed.**
- `require_batch_sizes(value, *, owner)` in `src/nnx/_validation.py` (built on FIX-021's `require_count`): `None` is the only automatic "one batch holding the complete split" sentinel; positive integral entries (NumPy integers normalized to `int`) are kept verbatim; zero, `False`, negatives, floats (`2.0` included), strings and NaN raise `ValueError` naming `batch_sizes[i] (split)`; a list, a bare value or a wrong-length tuple raises with the 3-tuple contract.
- `NNDataset`, `NNTabularDataset`, `NNPreferenceDataset`, `NNGraphDataset` call it first — before the dataset factory, `tokenizer.encode` or any RNG-consuming split — and resolve with `is None` instead of truthiness. Empty-split conventions (placeholder `1` + `None` loader), the vision test loader and `sampler="full"`'s rejection of explicit sizes (now checked after validation, so an explicit zero is rejected as a bad count rather than accepted as "disable the split") are unchanged.
- Example 22: `build_preference_dataset(...)` (used by `main()`) plus the bounded offline `dpo_sample_batch_sizes()` helper (counting stub tokenizer, no `lm` extra, no training), registered in the smoke harness. Docs: wrapper docstrings (`NNGraphDataset` gains its first class docstring), `docs/quickstart.md` §2.5 "Batch sizes", `docs/dpo.md`, `docs/concepts.md` §2.3 pointer, `examples/README.md`, `CHANGELOG.md [Unreleased]`, regenerated `docs/api.md`.

**Acceptance criteria → tests** (`tests/test_dataset_batch_sizes.py`, 124 tests + the sample helper):
- Zero / False / negative / fractional / boolean / string / NaN entries fail with split/index context in all four wrappers with zero factory / tokenizer calls — `test_invalid_entries_fail_with_split_context_before_any_dataset_work` (96 cases).
- Malformed lengths, lists and bare values fail with the 3-tuple contract before dataset work — `test_malformed_tuples_fail_with_the_tuple_contract_named`.
- `None` → full split (27/3/12, 2/1/1, 8/1/1, 4/2/2) with one train batch; `7` stays `7` in metadata while iteration yields one 2-row batch; `np.int64(5)` → `int` — `test_none_resolves_to_the_full_split_and_positive_sizes_are_kept_verbatim`, `test_ticket_oracle_tabular`.
- Empty-split conventions intact with explicit sizes; full-graph mode still rejects explicit sizes — `test_tabular_empty_split_conventions_are_unchanged_with_explicit_sizes`, `test_full_graph_sampler_still_rejects_explicit_sizes_after_validation`.
- Training cadence follows the resolved loader (1 step/epoch by default, 4 with `(3, None, None)` on ten rows) — `test_training_cadence_follows_the_resolved_train_loader`.
- DPO sample helper: invalid tuples → zero `encode` calls; `None` vs `(4, 2, None)` → 1 vs 4 train batches — `dpo_sample_batch_sizes`.

**Checks (both PRs, local + CI green: lint-and-test 3.10–3.14, pip-audit, CodeQL, docs build):**
```
uv run --frozen pytest -q                                    → 2132 passed, 1 skipped, 2 deselected (libcairo docs-projection tests, CI runs them)
uv run --frozen ruff check src/ tests/ examples/ scripts/    → All checks passed
uv run --frozen ruff format --check …                        → 257 files already formatted
uv run --frozen pyright --warnings                           → 0 errors, 0 warnings
build_api_reference / build_docs / mkdocs build --strict / build_wiki --check / projection tests → pass
python examples/22_dpo_synthetic_preferences.py              → completes
```
Not verifiable locally: `NeighborLoader` iteration (needs pyg-lib / torch-sparse); graph tests check construction, resolved sizes and loader length, as the existing graph suite does. Live site/wiki publication runs from the main push; not re-verified here.
